Mathlib/LinearAlgebra/Multilinear/DirectSum.lean

```diff
@@ -0,0 +1,88 @@
+/-
+Copyright (c) 2024 Sophie Morel. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sophie Morel
+-/
+import Mathlib.Algebra.DirectSum.Module
+import Mathlib.LinearAlgebra.Multilinear.DFinsupp
+
+/-!
+# Multilinear maps from direct sums
+
+This file describes multilinear maps on direct sums.
+
+## Main results
+
+* `MultilinearMap.fromDirectSumEquiv` : If 'ι` is a `Fintype`, `κ i` is a family of types
+indexed by `ι` and we are given a `R`-module `M i j` for every `i : ι` and `j : κ i`, this is
+the linear equivalence between `Π p : (i : ι) → κ i, MultilinearMap R (fun i ↦ M i (p i)) M'` and
+`MultilinearMap R (fun i ↦ ⨁ j : κ i, M i j) M'`.
+-/
+
+namespace MultilinearMap
+
+open DirectSum LinearMap Function
+
+variable (R : Type*) [CommSemiring R]
+variable {ι : Type*} [Fintype ι] [DecidableEq ι]
+variable (κ : ι → Type*) [(i : ι) → DecidableEq (κ i)]
+variable {M : (i : ι) → κ i → Type*} {M' : Type*}
+variable [Π i (j : κ i), AddCommMonoid (M i j)] [AddCommMonoid M']
+variable [Π i (j : κ i), Module R (M i j)] [Module R M']
+
+/-- Two multilinear maps from direct sums are equal if they agree on the generators. -/
+@[ext]
+theorem directSum_ext ⦃f g : MultilinearMap R (fun i ↦ ⨁ j : κ i, M i j) M'⦄
+    (h : ∀ p : Π i, κ i,
+      f.compLinearMap (fun i => DirectSum.lof R _ _ (p i)) =
+      g.compLinearMap (fun i => DirectSum.lof R _ _ (p i))) : f = g :=
+  dfinsupp_ext h
+
+/-- The linear equivalence between families indexed by `p : Π i : ι, κ i` of multilinear maps
+on the `fun i ↦ M i (p i)` and the space of multilinear map on `fun i ↦ ⨁ j : κ i, M i j`.-/
+def fromDirectSumEquiv :
+    ((p : Π i, κ i) → MultilinearMap R (fun i ↦ M i (p i)) M') ≃ₗ[R]
+    MultilinearMap R (fun i ↦ ⨁ j : κ i, M i j) M' :=
+  LinearEquiv.ofLinear
+    ((DirectSum.toModule R _ _ fun _ ↦ .id).compMultilinearMapₗ R ∘ₗ MultilinearMap.dfinsuppFamilyₗ)
+    (LinearMap.pi fun p ↦ MultilinearMap.compLinearMapₗ fun i ↦ DirectSum.lof R (κ i) _ (p i))
+    (by
+      ext f x
+      dsimp
+      erw [MultilinearMap.dfinsuppFamily_single, DirectSum.toModule_lof]
```

Comment:
These `erw`s are my fault, and are because we are abusing defeq by using `MultilinearMap.dfinsuppFamilyₗ` instead of duplicating the construction for `directSum`. Probably we should spin out a PR to do that, where all the defs and lemmas are one-liners in terms of the `dfinsupp` versions.